include/sparrow/array/array_data.hpp

```diff
@@ -52,6 +52,42 @@ namespace sparrow
         value_ptr<array_data> dictionary;
     };
 
+    data_descriptor type_descriptor(const array_data& data);
+    array_data::length_type length(const array_data& data);
+    std::int64_t offset(const array_data& data);
+
+    array_data::bitmap_type& bitmap(array_data& data);
+    const array_data::bitmap_type& bitmap(const array_data& data);
+
+    std::size_t buffers_size(const array_data& data);
+    array_data::buffer_type& buffer_at(array_data& data, std::size_t i);
+    const array_data::buffer_type& buffer_at(const array_data& data, std::size_t i);
+
+    std::size_t child_data_size(const array_data& data);
+    array_data& child_data_at(array_data& data, std::size_t i);
+    const array_data& child_data(const array_data& data, std::size_t i);
+
+    value_ptr<array_data>& dictionary(array_data& data);
+    const value_ptr<array_data>& dictionary(const array_data& data);
+
+    /**
+     * Concept for a structure that can be used as a data storage in the layout and the
+     * typed_array class.
+     */
+    template <class T>
+    concept data_storage = requires(T t, std::size_t i)
+    {
+        type_descriptor(t);
```

Comment:
I think we can add `-> expected-return-type` on every function call here? Maybe for some the return type can just be a concept.

Reply:
For most of the functions, yes. For the buffer and bitmap accessors, we cannot really because the return types are really different.

Reply:
Can at least a concept be specified? if not nevermind

Reply:
We can require a random_access_range, it's not complete but at least covers most of our usages.